File: static/Identifiable/Identifiable.cpp
#include "Identifiable.h"

Identifiable::Identifiable() {
    total += 1;
    newId += 1;
    this->id = newId;
}

Identifiable::Identifiable(const Identifiable& copy) {
    total += 1;
    newId += 1;
    this->id = newId;
}

Identifiable& Identifiable::operator=(const Identifiable& copy) {
    return *this;
}

Identifiable::~Identifiable() {
    total -= 1;
}

int Identifiable::getId() const {
    return this->id;
}

int Identifiable::getTotal() {
    return total;
}

int Identifiable::total = 0;
int Identifiable::newId = 0;

File: PHP/Classes/Identifiable/Identifiable.php
<?php
    class Identifiable implements Countable {
        private $id;
        private static $newId = 1;
        private static $total = 0;
        
        public function __construct() {
            $this->id = self::$newId;
            self::$newId += 1;
            self::$total += 1;
        }
        
        public function __destruct() {
            self::$total -= 1;
        }
        
        public function __clone() {
            $this->id = self::$newId;
            self::$total += 1;
            self::$newId += 1;
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                if ( $name == "id" ) {
                    return $this->$name;
                } else {
                    return self::$$name;
                }
            }
            throw new Exception("Atribute error: atribute $name not found");
        }
        
        public function count() {
            return self::$total;
        }
    }
?>
